react/features/local-recording/controller/RecordingController.js

```diff
@@ -0,0 +1,493 @@
+     * retrieved after they ended.
+     *
+     * @private
+     */
+    _adapters = {};
+
+    /**
+     * The {@code JitsiConference} instance.
+     *
+     * @private
+     */
+    _conference: * = null;
+
+    /**
+     * Current recording session token.
+     * Session token is a number generated by the moderator, to ensure every
+     * client is in the same recording state.
+     *
+     * @private
+     */
+    _currentSessionToken: number = -1;
+
+    /**
+     * Current state of {@code RecordingController}.
+     *
+     * @private
+     */
+    _state = ControllerState.IDLE;
+
+    /**
+     * Current recording format. This will be in effect from the next
+     * recording session, i.e., if this value is changed during an on-going
+     * recording session, that on-going session will not use the new format.
+     *
+     * @private
+     */
+    _format = DEFAULT_RECORDING_FORMAT;
+
+    /**
+     * Whether or not the {@code RecordingController} has registered for
+     * XMPP events. Prevents initialization from happening multiple times.
+     *
+     * @private
+     */
+    _registered = false;
+
+    /**
+     * FIXME: callback function for the {@code RecordingController} to notify
+     * UI it wants to display a notice. Keeps {@code RecordingController}
+     * decoupled from UI.
+     */
+    onNotify: ?(string) => void;
+
+    /**
+     * FIXME: callback function for the {@code RecordingController} to notify
+     * UI it wants to display a warning. Keeps {@code RecordingController}
+     * decoupled from UI.
+     */
+    onWarning: ?(string) => void;
+
+    /**
+     * FIXME: callback function for the {@code RecordingController} to notify
+     * UI that the local recording state has changed.
+     */
+    onStateChanged: ?(boolean) => void;
+
+    /**
+     * Constructor.
+     *
+     * @returns {void}
+     */
+    constructor() {
+        this._updateStats = this._updateStats.bind(this);
+        this._onStartCommand = this._onStartCommand.bind(this);
+        this._onStopCommand = this._onStopCommand.bind(this);
+        this._doStartRecording = this._doStartRecording.bind(this);
+        this._doStopRecording = this._doStopRecording.bind(this);
+        this.registerEvents = this.registerEvents.bind(this);
+        this.getParticipantsStats = this.getParticipantsStats.bind(this);
+    }
+
+    registerEvents: () => void;
+
+    /**
+     * Registers listeners for XMPP events.
+     *
+     * @param {JitsiConference} conference - {@code JitsiConference} instance.
+     * @returns {void}
+     */
+    registerEvents(conference: Object) {
+        if (!this._registered) {
+            this._conference = conference;
+            if (this._conference) {
+                this._conference
+                    .addCommandListener(COMMAND_STOP, this._onStopCommand);
+                this._conference
+                    .addCommandListener(COMMAND_START, this._onStartCommand);
```

Comment:
You mentioned test case about a reload. I assume that's the same issue around a new participant joining and potentially missing a start command. Is that correct?

Reply:
Yeah. When I register for XMPP events in the middleware, after `CONFERENCE_JOINED`, the start command (in remote presence) is already missed. Current workaround is let each client send a probe (sendCommandOnce) to force the moderator to resend the command. :(

Reply:
https://github.com/jitsi/jitsi-meet/pull/3223/commits/5f48400c69c531e5818408215fef762df892434e

Reply:
Another issue I observed is when the moderator reloads, the command will be removed from the presence. Is there a way to keep START_COMMAND in presence regardless of whether its origin sender is connected? START_COMMAND/STOP_COMMAND should be some kind of "persistent state" that belongs to the room.

Reply:
We've been thinking about the same problem in other situations as well. Let's discuss it when we talk. I think it's OK to merge it this way.